src/common/imagpng.cpp

```diff
@@ -394,49 +435,47 @@ wxPNGHandler::LoadFile(wxImage *image,
     }
 
 
-    png_destroy_read_struct( &png_ptr, &info_ptr, (png_infopp) NULL );
-
     // loaded successfully, now init wxImage with this data
-    CopyDataFromPNG(image, lines, width, height, color_type);
+    CopyDataFromPNG(image, data.lines, width, height, color_type);
 
-    for ( i = 0; i < height; i++ )
-        free( lines[i] );
-    free( lines );
-    delete lines_p;
-
-    return true;
+    // This will indicate to the caller that loading succeeded.
+    data.ok = true;
 
 error:
-    if (verbose)
-    {
-       wxLogError(_("Couldn't load a PNG image - file is corrupted or not enough memory."));
-    }
-
-    if ( image->IsOk() )
-    {
-        image->Destroy();
-    }
+    // Note that we only get here if both png_ptr and info_ptr are valid,
+    // otherwise we just return from this function early.
+    png_destroy_read_struct( &png_ptr, &info_ptr, (png_infopp) NULL );
+}
 
-    if ( lines )
-    {
-        for ( unsigned int n = 0; n < height; n++ )
-            free( lines[n] );
+bool
+wxPNGHandler::LoadFile(wxImage *image,
+                       wxInputStream& stream,
+                       bool verbose,
+                       int WXUNUSED(index))
+{
+    wxPNGInfoStruct wxinfo;
+    wxinfo.verbose = verbose;
+    wxinfo.stream.in = &stream;
 
-        free( lines );
-    }
-    delete lines_p;
+    wxPNGImageData data;
+    DoLoadPNGFile(image, wxinfo, data);
 
-    if ( png_ptr )
+    if ( !data.ok )
     {
-        if ( info_ptr )
+        if (verbose)
+        {
+           wxLogError(_("Couldn't load a PNG image - file is corrupted or not enough memory."));
+        }
+
+        if ( image->IsOk() )
         {
-            png_destroy_read_struct( &png_ptr, &info_ptr, (png_infopp) NULL );
-            free(info_ptr);
```

Comment:
I can't find this line in the new version, was it removed intentionally?

Reply:
Yes, I think this is not actually needed because passing `info_ptr` to `png_destroy_read_struct()` takes care of freeing it too.